File: preppi_slim/SCR/run_pipeline.pl
#!/usr/bin/perl

use MODS::Globals;
use strict;
use warnings;
use Getopt::Std;
use MODS::Methods;

$|++;

if(@ARGV>4) {
  print STDERR "Usage:  run_pipeline.pl pipeline_name genome_name [-d]\n";
  exit(-1);
  }

my $pname=shift;
my $gname=shift;

my %opts;
getopts('dm:',\%opts);
my $debug='no';
$debug='yes' if $opts{d};

my $pipeline=new MODS::Pipeline(name=>$pname,gname=>$gname,debug=>$debug);
chdir $pipeline->{pipedir}
    or die "Could not change directory to $pipeline->{pipedir}: $!\n";
my @steps=$pipeline->steps();
my $genome=new MODS::Genome(gname=>$gname);

$pipeline->stage("Pipeline started.");
sleep(5);

my $start=time();
my @running;
@running=`cat $pipeline->{pipedir}/running` if -e "$pipeline->{pipedir}/running";
`rm -rf $pipeline->{pipedir}/running` if -e "$pipeline->{pipedir}/running";
for(@running) { chop; }

my $n=0;
my %jobstat;
if(-e "$pipeline->{pipedir}/$pname.jstat") {
    my @jstat=`cat $pipeline->{pipedir}/$pname.jstat`; 
    for(@jstat) {
        chomp;
        my ($jid,$jinfo)=split(/\t/);
        $pipeline->stage("Reading jobs status: $_.") if not ($n++)%100;
        $jobstat{$jid}=$jinfo;
    }
}

my @jlist;
my @tgts=();
@tgts=`cat targets.lst` if -e "targets.lst";
@tgts=$genome->get_target_list() if not scalar(@tgts);
$n=0;
for(@tgts) { 
    $pipeline->stage("Setting up job queue: $_.") if not ($n++)%100;
    chomp; 
    foreach my $s (@steps) { 
        my $jid="$_:$s";
        my $tm=$pipeline->time();
        $jobstat{$jid}="($tm init)" if not defined $jobstat{$jid};
        push(@jlist,$jid) if not $jobstat{$jid}=~/fail|complete/;  
    }
}


my ($jtot,$nrun,$navail,$jactive)=(0,0,0,0);
my $ccount=0;
my $lasttgt=0;
$lasttgt=`cat $pname.tlast` if -e "$pname.tlast";
my %avail=();
 
if(defined $opts{m}) {
    my $end=$opts{m};
    @jlist=@jlist[0 .. $end];
}

while(1) {

    my $i;
    my $jfail;
    my $updatetm=time();
    for($i=$lasttgt;$i<scalar(@jlist);$i++) {
        last if -e "$pname.pause";
        last if -e "$pname.kill";
        last if -e "$pname.update";
        chomp $jlist[$i];
        my $jid=$jlist[$i];
        if(not $ccount%100) {
            $jactive=0;
            $jfail=0;
            for(@running) {
                my ($jname)=split;  
                if(-d "$jname.active") { 
                    `ls -f $jname.active | grep RUN | wc -l`=~/(\d+)/; $jactive+=$1; 
                    `ls -f $jname.active | grep FAIL | wc -l`=~/(\d+)/; $jfail+=$1;
                    last if $jfail>100;
                }
            }
            $pipeline->stage("Checking job status: $jid, $ccount jobs checked in current block, $jactive active jobs");
        }
        if(not $jid=~/:/) { print STDERR "jid error target $i"; print scalar(@jlist); print "\n"; exit(0); }
        my ($gid,$sname)=split(/:/,$jid);
	# last if $ccount++>2000;
	last if $ccount++>$MAX_ARRAY_VAL;
        my $jinfo="";
        my $tm=$pipeline->time();
        $jobstat{$jid}="($tm init)" if not defined $jobstat{$jid};
        $jinfo=$jobstat{$jid};
        $pipeline->status($gid,$sname,\%jobstat);
        print STDERR "$jid $jobstat{$jid}\n" if $pipeline->{debug} eq 'yes' and $jinfo ne $jobstat{$jid}; 
        if($jobstat{$jid}=~/complete|skip|fail/) { splice(@jlist,$i,1); $i--; next; }
        next if not $jobstat{$jid}=~/ready (\d+) jobs/ or $jobstat{$jid}=~/queued/;
        $jobstat{$jid}.="($tm queued)";
        `echo -e "$gid\t$1" >> $sname.tgt`;
    }

    $ccount=0;
    my $n=0;
    if(-e "$pname.update") {
        for(@running) {
            my ($jname,$nj)=split;
            my $sname=substr($jname,0,length($jname)-10);
            my @tms=`ls -f $jname.active | grep TM | cut -c 4-9;`;
            foreach my $gid (@tms) { 
                chop $gid;
                $pipeline->stage("Checking and saving job status: $gid $sname.") if ($n++)%100==0;
                $pipeline->status($gid,$sname,\%jobstat);
            }
        }
    }

    if($jfail) {
        for(@running) {
            my ($jname,$nj)=split;
            my $sname=substr($jname,0,length($jname)-10);
            my @failj=`ls -f $jname.active | grep FAIL | cut -c 6-100`;
            foreach my $gid (@failj) { 
                chop $gid;
                $pipeline->status($gid,$sname,\%jobstat);
                `rm -r $jname.active/FAIL.$gid`;
            }
        }
    }

    open(CPL,">$pipeline->{pipedir}/$pname.jstat");
    foreach my $k (keys %jobstat) { print CPL "$k\t$jobstat{$k}\n"; }
    close CPL;

    if(-e "$pname.update") { `cp $pname.jstat $pname.jstat.tmp; rm $pname.update`; }
    if(-e "$pname.kill") { $pipeline->stage("Pipeline was killed"); `rm -rf $pname.kill`; exit(0); }
    if(-e "$pname.pause") { `touch $pname.paused`; while(-e "$pname.pause") { sleep(1); } }
    if(time()-$updatetm>30) { `cp $pname.jstat $pname.jstat.tmp`; }
   
    $pipeline->stage("Checking cpu/space usage and submitting jobs.");
    
    $navail=0;
    for(@steps) {
        next if not -e "$_.tgt";
        `wc -l $_.tgt`=~/(\d+)/;
        $navail++ if $1!=0;
    }

    $pipeline->check_active();

    $nrun=0; $jtot=0;
    my @checkrun=@running;
    @running=();
    my %nstep=();
    for(@checkrun) {
        my ($jname,$j)=split;
        my $js=substr($jname,0,length($jname)-10);
        next if not -e "$jname";
        my @jobdat=`squeue --job=$j --noheader --format="%.j" 2> /dev/null`;
        next if not @jobdat;
        push(@running,$_);
        $nstep{$js}++ if not -e "$jname.active/LAST";
        $nrun++ if not -e "$jname.active/LAST";
    }

    for(@steps) {
        my $sname=$_;
        `wc -l $_.tgt`=~/(\d+)/;
        next if $1==0;
        $nstep{$sname}=0 if not defined $nstep{$sname};
        if((($nstep{$sname}<3 and $navail>1) or $navail==1) and $nrun<5) {
            push(@running,$pipeline->submit_block($sname,\%jobstat)); 
            open(RUNFH,">$pipeline->{pipedir}/running");
            for(@running) { print RUNFH "$_\n" }
            close RUNFH;
        } else {
            print STDERR "$sname\nnstep: $nstep{$sname} navail: $navail nrun: $nrun jtot: $jtot\n";
        }
    }

    $lasttgt=$i+1>=scalar(@jlist) ? 0 : $i;

    my $activej=scalar(@jlist);
    print STDERR "Remaining jobs: $activej\n";

    if(not scalar(@jlist)) {
        `cp $pname.jstat $pname.jstat.tmp`;
        my @failed=grep { $jobstat{$_}=~/fail/ } keys %jobstat;
        if(@failed) {
            $pipeline->stage("Pipeline failed: ".scalar(@failed)." job(s) failed.");
            exit(1);
        }
        $pipeline->stage("Pipeline complete.");
        $pipeline->mail(); 
        exit(0);
    }

    my $tm=time()-$start;
    print STDERR "Time: $tm\n";

    if((time()-$start)>10000) { 
        print STDERR "Time limit exceeded. Resubmitting.\n";
        $pipeline->qsub(); 
        `echo $lasttgt > $pname.tlast`;
        exit(0); 
    } 
    sleep(1);
}
   

File: preppi_slim/SCR/run_pipeline.py
#!/usr/bin/env python3
import argparse
import os
import re
import subprocess
import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from MODS.Pipeline import Pipeline


def queued_count(path):
    return sum(1 for line in path.read_text().splitlines() if line.strip())


def slurm_active(job_id):
    result = subprocess.run(
        ["squeue", "-h", "-j", str(job_id)],
        text=True, capture_output=True, check=False,
    )
    return bool(result.stdout.strip())


def save_status(path, status):
    temporary = path.with_suffix(path.suffix + ".new")
    temporary.write_text("".join(f"{key}\t{value}\n" for key, value in sorted(status.items())))
    temporary.replace(path)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("pipeline")
    parser.add_argument("genome")
    parser.add_argument("-d", "--debug", action="store_true")
    parser.add_argument("-m", "--maximum", type=int)
    args = parser.parse_args()
    pipeline = Pipeline(
        name=args.pipeline, gname=args.genome,
        debug="yes" if args.debug else "no",
    )
    os.chdir(pipeline.pipedir)
    steps = pipeline.steps()
    targets_file = Path("targets.lst")
    targets = (
        [line.strip() for line in targets_file.read_text().splitlines() if line.strip()]
        if targets_file.exists() else pipeline.genome.get_target_list()
    )
    pipeline.initialize_status(targets)
    status_path = Path(pipeline.pipedir, f"{pipeline.name}.jstat")
    status = {}
    for line in status_path.read_text().splitlines():
        if "\t" in line:
            key, info = line.split("\t", 1)
            status[key] = info
    pending = []
    reported_state = {
        f"{target}:{step}": "pending"
        for target in targets
        for step in steps
    }
    for target in targets:
        for step in steps:
            key = f"{target}:{step}"
            status.setdefault(key, f"({pipeline.time()} init)")
            if not re.search(r"fail|complete", status[key]):
                pending.append(key)
    if args.maximum is not None:
        pending = pending[:args.maximum]
    running_path = Path(pipeline.pipedir, "running")
    running = [line for line in running_path.read_text().splitlines() if line.strip()]
    pipeline.stage("Pipeline started.")
    started = time.time()

    while True:
        for key in list(pending):
            target, step = key.split(":", 1)
            pipeline.status(target, step, status)
            info = status[key]
            semantic_state, _ = pipeline.controller_state(info)
            if semantic_state != reported_state.get(key):
                pipeline.record_status(target, step, info)
                reported_state[key] = semantic_state
            if re.search(r"complete|skip|fail", info):
                pending.remove(key)
                continue
            match = re.search(r"ready (\d+) jobs", info)
            if match and "queued" not in info:
                status[key] += f"({pipeline.time()} queued)"
                pipeline.record_status(target, step, status[key])
                reported_state[key] = "queued"
                with open(Path(pipeline.pipedir, f"{step}.tgt"), "a") as handle:
                    handle.write(f"{target}\t{match.group(1)}\n")

        save_status(status_path, status)
        retained = []
        running_steps = set()
        for line in running:
            fields = line.split()
            if len(fields) < 2:
                continue
            block, job_id = fields[0], fields[1]
            if slurm_active(job_id):
                retained.append(line)
                running_steps.add(re.sub(r"_[A-Za-z0-9]{5}\.tgt$", "", block))
            else:
                # afterany normally removes this. If its postprocessor failed,
                # clear the stale marker so status() records a failure.
                import shutil
                shutil.rmtree(
                    Path(pipeline.pipedir, block + ".active"),
                    ignore_errors=True,
                )
        running = retained

        available = [
            step for step in steps
            if queued_count(Path(pipeline.pipedir, f"{step}.tgt"))
        ]
        for step in available:
            if len(running) >= 5:
                break
            if step in running_steps and len(available) > 1:
                continue
            record = pipeline.submit_block(step, status)
            if record:
                running.append(record)
                running_steps.add(step)
                save_status(status_path, status)
        running_path.write_text("\n".join(running) + ("\n" if running else ""))

        print(f"Remaining jobs: {len(pending)}", file=os.sys.stderr)
        if not pending and not running:
            failed = [key for key, info in status.items() if "fail" in info]
            save_status(status_path, status)
            if failed:
                pipeline.stage(f"Pipeline failed: {len(failed)} job(s) failed.")
                raise SystemExit(1)
            pipeline.stage("Pipeline complete.")
            pipeline.cleanup_transient_state()
            return
        pipeline.stage("Checking cpu/space usage and submitting jobs.")
        if time.time() - started > 10000:
            pipeline.qsub()
            return
        time.sleep(1)


if __name__ == "__main__":
    main()
